### PR TITLE
AutocompleteManager Plugin

### PR DESCRIPTION
**Pull request by nathantannar4**
This adds an optional `AutocompleteManager` that is a plugin to the `MessageInputBar`. With it, autocomplete can easily be added to look like the following:

simulator screen shot - iphone x - 2018-01-30 at 01 00 14

## Features 
- Autocomplete from any prefix
- Colored highlighting customizable to each prefix
- Easy to use

The Example has been updated to reflect using the `AutocompleteManager` .

I believe @cwalo has already been using this in a pre-release beta version I sent him.

### CODE REVIEW

File: Example/Sources/ConversationViewController.swift [hunk 1]
```diff
@@ -33,6 +33,15 @@ class ConversationViewController: MessagesViewController {
     var messageList: [MockMessage] = []
     
     var isTyping = false
+    
+    /// The object that manages autocomplete
+    open lazy var autocompleteManager: AutocompleteManager = { [unowned self] in
```
Comment (SD10) [inline Example/Sources/ConversationViewController.swift]: I hate using unowned 😱

File: Sources/Protocols/AutocompleteManagerDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ */
+
+import UIKit
+
+/// AutocompleteManagerDataSource is a protocol that passes data to the AutocompleteManager
+public protocol AutocompleteManagerDataSource: class {
```
Comment (SD10) [inline Sources/Protocols/AutocompleteManagerDataSource.swift]: Let's use `AnyObject` protocol

**Review by SD10 — COMMENTED** [review]
Great work @nathantannar4, I just have a few concerns:

- We're adding a lot of complexity here, a table view, more cells, more protocols
- We still haven't finished the v1.0 requirements
- Nothing is unit tested and I'd like to stop accepting untested features after 1.0. My hopes are after the layout API is refactored to be open, we can increase test coverage
- This is almost better suited as a potential plugin to the library after version 1.0. I think it's too early to decide

**Comment by cwalo** [comment]
LGTM 👍 

Indeed, I've been using a beta version in production, with some modifications - specifically, handling spaces within autocomplete text (we don't use usernames and the organizations using our app are large, so first name alone isn't sufficient). @nathantannar4 let me know if you want that considered and I'll try to put together a PR for it. I was in a rush to get @mentions working all together (tokenization/detokenization), so it's in a hacky state at the moment.

I tend to agree with @SD10 that we should get MessageKit to 1.0 before we add anything else, though. Additionally, we might want to consider marrying this to a feature that parses @mentions in message text cells - this can be client/user specific though, so it might be difficult to handle every case (see above). The work required for our case was not pleasant, haha.

**Comment by Sherlouk** [comment]
Looks pretty neat but I would've quite liked to have seen support for async responses to autocomplete similar to the way we handle it in [`MessageVC`](https://github.com/GitHawkApp/MessageViewController) as I can definitely see the need to use APIs to load certain information here.

Also not a huge fan of passing 3 parameter tuples around, doesn't look fairly nice and makes the function definitions lengthy and in places hard to read. I would recommend switching these out for a struct (or at bare minimum use a type alias)

Definitely re-iterating previous remarks though this is a pretty substantial change without being sufficiently tested (we've definitely seen our autocomplete fail a few times on GH, though we're starting to get some tests down)

**Comment by cwalo** [comment]
@Sherlouk good call on the tuples.

**Comment by nathantannar4** [comment]
Thanks for everyones feedback. I am working on:
1. No more tuples
2. A way to handle autocompletes that have spaces
3. More flexibility
4. Better API

**Comment by nathantannar4** [comment]
Made some revisions

**Comment by nathantannar4** [comment]
Ok this is ready for a final review

**Comment by SD10** [comment]
Still has to be blocked by MessageKit 1.0 😕

**Comment by SD10** [comment]
Closing this one

**Pull request closed without merging**